# Classificação Redes Neurais - Wisconsin Diagnostic Breast Cancer

## Descrição da atividade

- Nesta lista vamos fazer a predição para saber se um tumor de mama é benigno ou maligno a partir do dataset "Wisconsin Diagnostic Breast Cancer (WDBC)".
- Será utilizado 70% dos dados para treino e outros 30% para teste.
- Vamos criar modelos de aprendizado de máquina usando os seguintes algoritmos:
  - Perceptron
  - Adaline com Gradiente Descendente
  - Adaline com Gradiente Descentente Estocástico
  - Adaline com Gradiente Descendente Estocástico usando mini-batches de 20 elementos (mesmo do item anterior, mas agora com mini-batch).
- Os resultados obtidos entre os algoritmos acima usando a métrica acurácia serão comparados no final desta atividade.

## Ajustando os dados

In [285]:
from sklearn.datasets import load_breast_cancer
from sklearn.model_selection import train_test_split
from sklearn.preprocessing import StandardScaler

In [286]:
data = load_breast_cancer()
X = data.data
y = data.target

# Dividindo os dados:
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.3, stratify=y)

# Normalizando os dados:
normalizacao = StandardScaler()
normalizacao.fit(X_train)

X_train_std = normalizacao.transform(X_train)
X_test_std = normalizacao.transform(X_test)

## Importações Gerais

In [287]:
from sklearn.metrics import accuracy_score
import numpy as np

## Perceptron

In [288]:
from sklearn.linear_model import Perceptron

In [370]:
perc = Perceptron(random_state=5)

perc.fit(X_train_std, y_train)
y_pred = perc.predict(X_test_std)

print("Accuracy: {0:.2f}%".format((accuracy_score(y_test, y_pred)) * 100))

Accuracy: 97.08%


## Adaline com Gradiente Descendente

In [344]:
class Adaline(object):
    def __init__(self, eta = 0.001, epoch = 100):
        self.eta = eta
        self.epoch = epoch

    def fit(self, X, y):
        np.random.seed(16)
        self.weight_ = np.random.uniform(-1, 1, X.shape[1] + 1)
        self.error_ = []
        
        cost = 0
        for _ in range(self.epoch):
            
            output = self.activation_function(X)
            error = y - output
            
            self.weight_[0] += self.eta * sum(error)
            self.weight_[1:] += self.eta * X.T.dot(error)
            
            cost = 1./2 * sum((error**2))
            self.error_.append(cost)
            
        return self

    def net_input(self, X):
        """Calculo da entrada z"""
        return np.dot(X, self.weight_[1:]) + self.weight_[0]
    def activation_function(self, X):
        """Calculo da saida da funcao g(z)"""
        return self.net_input(X)
    def predict(self, X):
        """Retornar valores binaros 0 ou 1"""
        return np.where(self.activation_function(X) >= 0.0, 1, -1)

In [345]:
adGD = Adaline()

adGD.fit(X_train_std, y_train)
y_pred = adGD.predict(X_test_std)

print("Accuracy: {0:.2f}%".format((accuracy_score(y_test, y_pred)) * 100))

Accuracy: 59.06%


## Adaline com Gradiente Descentente Estocástico

In [292]:
from sklearn.linear_model import SGDClassifier

In [371]:
adSGD = SGDClassifier(random_state=7)

adSGD.fit(X_train_std, y_train)
y_pred = adSGD.predict(X_test_std)

print("Accuracy: {0:.2f}%".format((accuracy_score(y_test, y_pred)) * 100))

Accuracy: 98.25%


## Adaline com Gradiente Descendente Estocástico (mini-batches de 20 elementos)

In [380]:
adSGD = SGDClassifier(random_state=7)

for i in range(0, len(X_train_std), 20):
    X_batch = X_train_std[i:i + 20]
    y_batch = y_train[i:i + 20]
    adSGD.partial_fit(X_batch, y_batch, classes=np.unique(y_train))

y_pred = adSGD.predict(X_test_std)

print("Accuracy: {0:.2f}%".format((accuracy_score(y_test, y_pred)) * 100))

Accuracy: 97.08%


## Resultados obtidos

|Algoritmos|Acurácia|
|----------|--------|
|Perceptron|97.08%|
|Adaline com Gradiente Descendente|59.06%|
|Adaline com Gradiente Descentente Estocástico|98.25%|
|Adaline com Gradiente Descendente Estocástico (mini-batches de 20 elementos)|97.08%|